android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/util/platforms/android/AndroidPlatform.java

```diff
@@ -0,0 +1,727 @@
+                    principalName(groupName.get()).principalIdentifier(groupId.get()).build();
+        }
+        return CURRENT_USER;
+    }
+
+    private AndroidUserAttributes lookupUser(String user) throws IOException {
+        if (Utils.isEmpty(user)) {
+            throw new IOException("No user to lookup");
+        }
+        boolean isNumeric = user.chars().allMatch(Character::isDigit);
+        AndroidUserAttributes.AndroidUserAttributesBuilder builder = AndroidUserAttributes.builder();
+
+        if (isNumeric) {
+            builder.principalIdentifier(user);
+        }
+        builder.principalName(user);
+        Optional<String> id = id(user, AndroidPlatform.IdOption.User, isNumeric);
+        if (id.isPresent()) {
+            if (isNumeric) {
+                builder.principalName(id.get());
+            } else {
+                builder.principalIdentifier(id.get());
+            }
+        } else {
+            if (!isNumeric) {
+                throw new IOException("Unrecognized user: " + user);
+            }
+        }
+
+        id(user, AndroidPlatform.IdOption.Group, false).ifPresent(s -> builder.primaryGid(Long.parseLong(s)));
+        return builder.build();
+    }
+
+    @SuppressWarnings("PMD.AssignmentInOperand")
+    private static AndroidGroupAttributes lookupGroup(String name) throws IOException {
+        if (Utils.isEmpty(name)) {
+            throw new IOException("No group to lookup");
+        }
+        if (CURRENT_USER_PRIMARY_GROUP.getClass()
+                .equals(name)) {
+            return CURRENT_USER_PRIMARY_GROUP;
+        } else {
+            throw new IOException("Non primary group is not supported");
+        }
+    }
+
+    @Override
+    public boolean userExists(String user) {
+        return true;
+    }
+
+    public AndroidUserAttributes lookupUserByIdentifier(String identifier) throws IOException {
+        return lookupUser(identifier);
+    }
+
+    @Override
+    public AndroidGroupAttributes lookupGroupByName(String group) throws IOException {
+        return lookupGroup(group);
+    }
+
+    @Override
+    public AndroidGroupAttributes lookupGroupByIdentifier(String identifier) throws IOException {
+        return lookupGroup(identifier);
+    }
+
+    @Override
+    public AndroidUserAttributes lookupCurrentUser() throws IOException {
+        return loadCurrentUser();
+    }
+
+    @Override
+    public Set<Integer> killProcessAndChildren(Process process, boolean force, Set<Integer> additionalPids,
+                                               UserDecorator decorator)
+            throws IOException, InterruptedException {
+        Integer ppid = -1;
+        try {
+            Field f = process.getClass().getDeclaredField("pid");
+            f.setAccessible(true);
+            ppid = (int) f.getLong(process);
+            f.setAccessible(false);
+        } catch (NoSuchFieldException | IllegalAccessException cause) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)